### PR TITLE
add glean to treeherder for measuring failure outcomes

### PR DESCRIPTION
**Pull request by jmaher**

### CODE REVIEW

**Comment by jmaher** [comment]
This patch adds glean to treeherder, specifically to measure pinned job annotation and new bugs filed.  These all take place from the detailsPanel as the common root.  

A few things feel hacky, I have tried to work around this.
1) the package.json commands with installing python + virtualenv; the tests/linters need the glean files, but they run on different environments, so it was sort of a hacky situation.  I am open to ideas here
2) jest.config.js changes seemed to be the only way to get the tests to pass
3) adding a property `newFailure` to the job isn't so hacky, but I worry if I overlooked an edge case
4) my use of origin.contains() is hacky, I would be happy to hear of other ways to determine if we are not in production

I have tested this often locally both with full docker and with just the front end, but using the staging (allizom.com) backend.  If you want to test this locally, you can see pings show up here:
https://debug-ping-preview.firebaseapp.com/pings/treeherder

File: ui/glean/pings.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+---
+# Schema
+$schema: moz://mozilla.org/schemas/glean/pings/2-0-0
+
+classification:
+  description: >
+    A ping to record treeherder classification data
+  include_client_id: false
```
Comment (Dexterp37) [inline ui/glean/pings.yaml:8]: Out of curiosity, why do you not want a client id?

File: ui/job-view/details/DetailsPanel.jsx [hunk 2]
```diff
@@ -20,6 +23,12 @@ import PinBoard from './PinBoard';
 import SummaryPanel from './summary/SummaryPanel';
 import TabsPanel from './tabs/TabsPanel';
 
+Glean.setLogPings(true);
+if (!origin.includes('treeherder.mozilla.org')) {
```
Comment (Dexterp37) [inline ui/job-view/details/DetailsPanel.jsx:27]: Does the frontend know about environment variables? If so, you could set something like `NODE_ENV='production'` and check that instead?

File: ui/job-view/details/PinBoard.jsx [hunk 3]
```diff
@@ -89,6 +112,13 @@ class PinBoard extends React.Component {
     }
     if (errorFree) {
       const jobs = Object.values(pinnedJobs);
+      // accumulate telemetry on failure classification type
+      // submit data when all jobs are counted
+      jobs.forEach((job) =>
+        gleanMetrics(failureClassificationId, job.newFailure),
+      );
+      classification.submit();
```
Comment (Dexterp37) [inline ui/job-view/details/PinBoard.jsx:120]: Can you expand a bit on what your intent is? Is there any chance for the user to close the page before this gets hit?
Another possibility would be to submit an event for every action of interest to you.

File: package.json [hunk 4]
```diff
@@ -114,20 +116,26 @@
     "webpack-merge": "5.8.0"
   },
   "scripts": {
-    "build": "node ./node_modules/webpack/bin/webpack.js --mode production",
+    "build": "yarn build:preglean && yarn build:gleanfiles && node ./node_modules/webpack/bin/webpack.js --mode production",
     "build:dev": "node ./node_modules/webpack/bin/webpack.js --mode development",
+    "build:preglean": "apt-get update && apt-get install python3-venv -y",
+    "build:pregleansudo": "sudo apt-get update && sudo apt-get install python3-venv",
```
[the remark this answers is not in the document]
Comment (jmaher) [reply package.json:122]: I tried adding python to the docker file originally and it appears we don't use a single docker file for all the different steps.  I don't really understand the whole setup.  It got funky because one of the tests ran on a system where I couldn't install without sudo and another I didn't have access to run sudo.

File: package.json [hunk 5]
```diff
@@ -114,20 +116,26 @@
...
+    "build:gleanfiles": "node ./node_modules/@mozilla/glean/dist/cli/cli.js translate ui/glean/metrics.yaml ui/glean/pings.yaml ui/glean/tags.yaml -f javascript -o ui/glean/generated",
+    "build:glean": "yarn build:pregleansudo && yarn build:gleanfiles",
```
[the remark this answers is not in the document]
Comment (jmaher) [reply package.json:124]: see above- I needed sudo to install the python package for a specific instance.

File: package.json [hunk 6]
```diff
@@ -114,20 +116,26 @@
...
     "codecov": "rm -rf coverage && node ./node_modules/codecov/bin/codecov",
+    "docs:glean": "node ./node_modules/@mozilla/glean/dist/cli/cli.js translate ui/glean/metrics.yaml ui/glean/pings.yaml ui/glean/tags.yaml -f markdown -o glean",
```
[the remark this answers is not in the document]
Comment (jmaher) [reply package.json:126]: in working with :bruno, we found that on the docker image the `glean` command was not available despite installing fresh, etc.  so we found that calling `node cli.js` (which is what `glean` maps to) works.

File: ui/job-view/details/DetailsPanel.jsx [hunk 7]
```diff
@@ -20,6 +23,12 @@ import PinBoard from './PinBoard';
...
+  Glean.setDebugViewTag('treeherder');
+}
+Glean.initialize('treeherder', true);
```
[the remark this answers is not in the document]
Comment (jmaher) [reply ui/job-view/details/DetailsPanel.jsx:30]: the goal is not to collect data about any person or machine, instead count # of transactions within treeherder.  The alternative is a database counter.  So I need to collect all instances of this, maybe I don't understand glean properly- if there is PII collected, is there a way to skip that or ignore it?

File: ui/glean/pings.yaml (hunk 1 above)
Comment (jmaher) [reply ui/glean/pings.yaml:8]: I only care about counting number of transactions and metrics from treeherder- I don't care who does it or where they are from or what type of computer they are on.

File: ui/job-view/details/DetailsPanel.jsx (hunk 2 above)
Comment (jmaher) [reply ui/job-view/details/DetailsPanel.jsx:27]: I couldn't figure that out, so instead I will as mentioned in a previous comment, comment out the debugview stuff.

File: ui/job-view/details/PinBoard.jsx (hunk 3 above)
Comment (jmaher) [reply ui/job-view/details/PinBoard.jsx:120]: the intent is when a transaction takes place to record it- I am doing the submit() at the end of the transaction so if the page is closed shortly thereafter we don't lose any metrics that might be submitted on a timer.

I need to learn more about what an event is vs the counter metric I am using.

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmaher) [reply package.json]: found a better way with the .circleci/config.yml !

File: ui/job-view/details/DetailsPanel.jsx [hunk 8]
```diff
@@ -20,6 +23,12 @@ import PinBoard from './PinBoard';
 import SummaryPanel from './summary/SummaryPanel';
 import TabsPanel from './tabs/TabsPanel';
 
+Glean.setLogPings(true);
```
[the remark this answers is not in the document]
Comment (jmaher) [reply ui/job-view/details/DetailsPanel.jsx:26]: I am not sure what it does, I saw it in an example

[the remark this answers is not in the document]
Comment (Dexterp37) [reply ui/job-view/details/DetailsPanel.jsx:26]: It prints the content of the pings to the console, it's useful for debugging

File: ui/job-view/details/DetailsPanel.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (Dexterp37) [reply ui/job-view/details/DetailsPanel.jsx:29]: This is not specifically about Glean itself, it's more a policy question: you'd have the same problem even if you used GA or any other data collection library. If data-stewards are happy with this, we should be fine!

File: ui/job-view/details/PinBoard.jsx (hunk 3 above)
Comment (Dexterp37) [reply ui/job-view/details/PinBoard.jsx]: I mean, that's fine: if you tested this and generates the data you're looking for, it's ok!

**Review by Archaeopteryx — APPROVED** [review]
